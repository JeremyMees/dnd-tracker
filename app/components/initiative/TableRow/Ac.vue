<script setup lang="ts">
import { reset } from '@formkit/core'
import { acFunctions } from '~/utils/dnd-helpers'

const props = defineProps<{
  item: InitiativeSheetRow
  sheet: InitiativeSheet | undefined
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
}>()

const { t } = useI18n()

const { add, remove, temp, override, overrideReset } = acFunctions

const popoverOpen = shallowRef<boolean>(false)
const selectedType = shallowRef<AcType>('remove')

const hasAc = computed(() => isDefined(props.item.ac) && isDefined(props.item.maxAc))

type AcType = 'add' | 'remove' | 'temp' | 'override' | 'override-reset'
interface AcForm { amount: number }

async function updateRow(row: Partial<InitiativeSheetRow>): Promise<void> {
  if (!props.sheet) return

  const index = getCurrentRowIndex(props.sheet, props.item.id)
  const rows = [...props.sheet.rows]

  if (index === -1 || !rows[index]) return

  rows[index] = { ...rows[index], ...row }

  await props.update({ rows })
  popoverOpen.value = false
}

async function updateBase(form: AcForm, node: FormNode): Promise<void> {
  node.clearErrors()

  try {
    if (!props.sheet) return

    const { amount } = sanitizeForm<AcForm>(form)
    const row = {
      ...props.item,
      maxAc: amount,
      maxAcOld: undefined,
      ac: amount,
    }

    await updateRow(row)
  }
  catch {
    reset('InitiativeRowAcBase')
    node.setErrors(t('general.error.text'))
  }
}

async function updateOverride(form: AcForm & { reset?: boolean }, node: FormNode): Promise<void> {
  node.clearErrors()

  try {
    if (!props.sheet) return

    const { amount, reset } = sanitizeForm<AcForm & { reset?: boolean }>(form)

    const row = reset || amount === props.item.maxAcOld
      ? handleAcChanges(props.item.maxAcOld ?? 0, 'override-reset')
      : handleAcChanges(amount, 'override')

    await updateRow(row)
  }
  catch {
    reset('InitiativeRowAcOverride')
    node.setErrors(t('general.error.text'))
  }
}

async function updateAc(form: AcForm, node: FormNode): Promise<void> {
  node.clearErrors()

  try {
    const selected = selectedType.value

    if (!props.sheet || !selected) return

    const { amount, type } = sanitizeForm<AcForm & { type?: AcType }>({ ...form, type: selected })

    const row = handleAcChanges(amount, type || 'remove')

    await updateRow(row)
  }
  catch {
    reset('InitiativeRowAcUpdate')
    node.setErrors(t('general.error.text'))
  }
}

function handleAcChanges(amount: number, type: AcType): InitiativeSheetRow {
  const row = { ...props.item }

  if (type === 'add') add(row, amount)
  else if (type === 'remove') remove(row, amount)
  else if (type === 'temp') temp(row, amount)
  else if (type === 'override') override(row, amount)
  else if (type === 'override-reset') overrideReset(row, amount)

  // when ac is an negative number change it to 0
  const resetNegative = props.sheet?.settings?.negative === false
  if (resetNegative && row.ac && row.ac < 0) row.ac = 0

  return row
}
</script>

<template>
  <div>
    <UiPopover v-model:open="popoverOpen">
      <UiPopoverTrigger as-child>
        <button
          data-test-trigger
          :class="{
            'bg-destructive/20 p-2 rounded-lg w-fit': isDefined(item.ac) && item.ac <= 0,
          }"
          class="flex flex-col gap-y-1"
        >
          <div class="flex items-center gap-x-1">
            <Icon
              v-if="!isDefined(item.ac) && item.type !== 'lair'"
              data-test-empty
              name="tabler:plus"
              class="size-5 min-w-5 text-foreground/10"
              aria-hidden="true"
            />
            <span
              v-else
              data-test-ac
              :class="{ 'text-destructive': isDefined(item.ac) && item.ac <= 0 }"
            >
              {{ item.ac }}
            </span>
            <span
              v-if="isDefined(item.ac) && item.tempAc"
              v-tippy="$t('general.temp')"
              data-test-temp
              class="text-warning text-xs"
            >
              +{{ item.tempAc }}
            </span>
          </div>
          <span
            v-if="item.maxAc !== item.ac"
            data-test-max
            class="text-2xs text-muted-foreground whitespace-nowrap"
          >
            {{ $t('general.max') }}: {{ item.maxAc }}
          </span>
        </button>
      </UiPopoverTrigger>
      <UiPopoverContent>
        <div
          v-if="isDefined(item.ac) && isDefined(item.maxAc)"
          class="flex flex-wrap gap-x-1 gap-y-2 pb-6 items-start justify-center"
        >
          <div class="p-2 rounded-lg space-y-4 min-w-[75px] bg-background text-center flex-1">
            <p class="font-bold text-muted-foreground text-xs">
              {{ $t('general.current') }}
            </p>
            <p
              class="head-2"
              :class="{ 'text-destructive': item.ac < 1 }"
            >
              {{ item.ac || 0 }}
            </p>
          </div>
          <div class="p-2 rounded-lg space-y-4 min-w-[75px] bg-background text-center flex-1">
            <p class="font-bold text-muted-foreground text-xs">
              {{ $t('general.max') }}
            </p>
            <div class="flex gap-1 items-start justify-center">
              <p
                class="head-2"
                :class="[!item.maxAcOld ? undefined : item.maxAcOld < item.maxAc ? 'text-success' : 'text-destructive']"
              >
                {{ item.maxAc || 0 }}
              </p>
              <p
                v-if="item.maxAcOld === 0 || item.maxAcOld"
                class="text-sm"
              >
                ({{ item.maxAcOld }})
              </p>
            </div>
          </div>
          <div class="p-2 rounded-lg space-y-4 min-w-[75px] bg-background text-center flex-1">
            <p class="font-bold text-muted-foreground text-xs">
              {{ $t('general.temp') }}
            </p>
            <p class="head-2">
              {{ item.tempAc || 0 }}
            </p>
          </div>
        </div>
        <FormKit
          v-if="hasAc"
          id="InitiativeRowAcUpdate"
          type="form"
          :actions="false"
          @submit="updateAc"
        >
          <FormDiceRollInput />
          <div class="flex items-center gap-x-2">
            <FormKit
              type="submit"
              input-class="$remove:btn-foreground btn-success flex items-center gap-x-2 px-2 border-2 text-sm"
              @click="selectedType = 'add'"
            >
              <Icon
                name="tabler:plus"
                class="size-4 min-w-4"
                aria-hidden="true"
              />
              {{ $t('actions.heal') }}
            </FormKit>
            <FormKit
              type="submit"
              input-class="$remove:btn-foreground btn-warning flex items-center gap-x-2 px-2 border-2 text-sm"
              @click="selectedType = 'temp'"
            >
              <Icon
                name="tabler:plus"
                class="size-4 min-w-4"
                aria-hidden="true"
              />
              {{ $t('actions.temp') }}
            </FormKit>
            <FormKit
              type="submit"
              input-class="$remove:btn-foreground btn-destructive flex items-center gap-x-2 px-2 border-2 text-sm"
              @click="selectedType = 'remove'"
            >
              <Icon
                name="tabler:sword"
                class="size-4 min-w-4"
                aria-hidden="true"
              />
              {{ $t('actions.damage') }}
            </FormKit>
          </div>
        </FormKit>
        <UiSeparator
          v-if="hasAc"
          class="my-6 bg-muted-foreground"
        />
        <FormKit
          id="InitiativeRowAcBase"
          type="form"
          :submit-label="$t('actions.save')"
          @submit="updateBase"
        >
          <FormKit
            type="number"
            name="amount"
            number
            :label="$t('components.inputs.baseFieldLabel', { field: 'AC' })"
            :help="$t('components.inputs.baseFieldHelp', { field: 'AC' })"
            validation="required|between:1,100|number"
          />
        </FormKit>
        <UiSeparator
          v-if="hasAc"
          class="my-6 bg-muted-foreground"
        />
        <FormKit
          v-if="hasAc"
          id="InitiativeRowAcOverride"
          type="form"
          :submit-label="$t('actions.save')"
          @submit="updateOverride"
        >
          <FormKit
            type="number"
            name="amount"
            number
            :label="$t('components.inputs.overrideFieldLabel', { field: 'AC' })"
            :help="$t('components.inputs.optionalFieldHelp', { field: 'AC' })"
            validation="required|between:1,100|number"
            :suffix-icon="item.maxAcOld ? 'tabler:player-skip-back' : undefined"
            @suffix-icon-click="(node: FormNode) => updateOverride({ reset: true, amount: item.maxAcOld || 0 }, node)"
          />
        </FormKit>
      </UiPopoverContent>
    </UiPopover>
  </div>
</template>
