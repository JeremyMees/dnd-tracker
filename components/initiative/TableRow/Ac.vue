<script setup lang="ts">
import { reset } from '@formkit/core'
import { acFunctions } from '~/utils/dnd-helpers'

const props = defineProps<{
  item: InitiativeSheetRow
  sheet: InitiativeSheet | undefined
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
}>()

const { t } = useI18n()

const { add, remove, temp, override, overrideReset } = acFunctions

const popoverOpen = ref<boolean>(false)

type AcType = 'add' | 'remove' | 'temp' | 'override' | 'override-reset'
interface AcForm { amount: number }

async function updateRow(row: Partial<InitiativeSheetRow>): Promise<void> {
  if (!props.sheet) return

  const index = getCurrentRowIndex(props.sheet, props.item.id)
  const rows = [...props.sheet.rows]

  if (index === -1) return

  rows[index] = { ...rows[index], ...row }

  await props.update({ rows })
  popoverOpen.value = false
}

async function updateOverride(form: AcForm & { reset?: boolean }, node: FormNode): Promise<void> {
  node.clearErrors()

  try {
    if (!props.sheet) return

    const { amount, reset } = sanitizeForm<AcForm & { reset?: boolean }>(form)

    const row = reset || amount === props.item.maxAcOld
      ? handleAcChanges(props.item.maxAcOld ?? 0, 'override-reset')
      : handleAcChanges(amount, 'override')

    await updateRow(row)
  }
  catch {
    reset('InitiativeRowAcOverride')
    node.setErrors(t('general.error.text'))
  }
}

async function updateAc(form: AcForm & { type?: AcType }, node: FormNode): Promise<void> {
  node.clearErrors()

  try {
    if (!props.sheet || !form.type) return

    const { amount, type } = sanitizeForm<AcForm & { type?: AcType }>(form)

    const row = handleAcChanges(amount, type || 'remove')

    await updateRow(row)
  }
  catch {
    reset('InitiativeRowAcUpdate')
    node.setErrors(t('general.error.text'))
  }
}

function handleAcChanges(amount: number, type: AcType): InitiativeSheetRow {
  const row = { ...props.item }

  if (type === 'add') add(row, amount)
  else if (type === 'remove') remove(row, amount)
  else if (type === 'temp') temp(row, amount)
  else if (type === 'override') override(row, amount)
  else if (type === 'override-reset') overrideReset(row, amount)

  // when ac is an negative number change it to 0
  if (row.ac && row.ac < 0) row.ac = 0

  return row
}
</script>

<template>
  <div>
    <UiPopover v-model:open="popoverOpen">
      <UiPopoverTrigger as-child>
        <button
          :class="{
            'bg-destructive/20 p-2 rounded-lg w-fit': item.ac !== null && item.ac === 0,
          }"
          class="flex flex-col gap-y-1"
        >
          <div class="flex items-center gap-x-1">
            <Icon
              v-if="item.ac === null"
              name="tabler:plus"
              class="size-5 min-w-5 text-foreground/10"
              aria-hidden="true"
            />
            <span
              v-else
              :class="{ 'text-destructive': item.ac === 0 }"
            >
              {{ item.ac }}
            </span>
            <span
              v-if="item.ac !== null && item.tempAc"
              v-tippy="$t('general.temp')"
              class="text-warning text-xs"
            >
              +{{ item.tempAc }}
            </span>
          </div>
          <span
            v-if="item.maxAc !== item.ac"
            class="body-extra-small text-muted-foreground whitespace-nowrap"
          >
            {{ $t('general.max') }}: {{ item.maxAc }}
          </span>
        </button>
      </UiPopoverTrigger>
      <UiPopoverContent>
        <div
          v-if="item?.ac !== undefined && item?.maxAc !== undefined"
          class="flex flex-wrap gap-x-1 gap-y-2 pb-6 items-start justify-center"
        >
          <div class="p-2 rounded-lg space-y-4 min-w-[75px] bg-background text-center flex-1">
            <p class="font-bold text-muted-foreground text-xs">
              {{ $t('general.current') }}
            </p>
            <p
              class="head-2"
              :class="{ 'text-destructive': item.ac < 1 }"
            >
              {{ item.ac || 0 }}
            </p>
          </div>
          <div class="p-2 rounded-lg space-y-4 min-w-[75px] bg-background text-center flex-1">
            <p class="font-bold text-muted-foreground text-xs">
              {{ $t('general.max') }}
            </p>
            <div class="flex gap-1 items-start justify-center">
              <p
                class="head-2"
                :class="[!item.maxAcOld ? undefined : item.maxAcOld < item.maxAc ? 'text-success' : 'text-destructive']"
              >
                {{ item.maxAc || 0 }}
              </p>
              <p
                v-if="item.maxHealthOld === 0 || item.maxHealthOld"
                class="body-small"
              >
                ({{ item.maxAcOld }})
              </p>
            </div>
          </div>
          <div class="p-2 rounded-lg space-y-4 min-w-[75px] bg-background text-center flex-1">
            <p class="font-bold text-muted-foreground text-xs">
              {{ $t('general.temp') }}
            </p>
            <p class="head-2">
              {{ item.tempAc || 0 }}
            </p>
          </div>
        </div>
        <FormKit
          id="InitiativeRowAcUpdate"
          type="form"
          :submit-label="$t('actions.save')"
          @submit="updateAc"
        >
          <FormDiceRollInput />
          <FormKit
            type="togglebuttons"
            name="type"
            value="remove"
            enforced
            :options="[
              { label: $t('actions.heal'), value: 'add', icon: 'tabler:heart' },
              { label: $t('actions.temp'), value: 'temp', icon: 'tabler:plus' },
              { label: $t('actions.damage'), value: 'remove', icon: 'tabler:sword' },
            ]"
            validation="required"
            input-class="$remove:p-2 $remove:border-4 $remove:aria-pressed:border-primary $remove:focus-within:border-primary group"
            option-class="$remove:w-10 $remove:h-10"
          >
            <template #default="{ option }">
              <div
                class="flex items-center gap-x-1 rounded-lg p-2 border-2 border-transparent"
                :class="{
                  'bg-destructive/20 group-aria-pressed:bg-destructive/40 group-aria-pressed:border-destructive': option.value === 'remove',
                  'bg-success/20 group-aria-pressed:bg-success/40 group-aria-pressed:border-success': option.value === 'add',
                  'bg-warning/20 group-aria-pressed:bg-warning/40 group-aria-pressed:border-warning': option.value === 'temp',
                }"
              >
                <Icon
                  :name="option.icon"
                  class="size-4 min-w-4 text-foreground"
                  aria-hidden="true"
                />
                <span class="text-xs text-muted-foreground group-aria-pressed:!text-foreground">
                  {{ option.label }}
                </span>
              </div>
            </template>
          </FormKit>
        </FormKit>
        <UiSeparator class="my-6 bg-muted-foreground" />
        <FormKit
          id="InitiativeRowAcOverride"
          type="form"
          :submit-label="$t('actions.save')"
          @submit="updateOverride"
        >
          <FormKit
            type="number"
            name="amount"
            number
            :label="$t('components.inputs.overrideFieldLabel', { field: 'AC' })"
            :help="$t('components.inputs.optionalFieldHelp', { field: 'AC' })"
            validation="required|between:1,100|number"
            :suffix-icon="item.maxAcOld ? 'tabler:player-skip-back' : undefined"
            @suffix-icon-click="(node: FormNode) => updateOverride({ reset: true, amount: item.maxAcOld || 0 }, node)"
          />
        </FormKit>
      </UiPopoverContent>
    </UiPopover>
  </div>
</template>
