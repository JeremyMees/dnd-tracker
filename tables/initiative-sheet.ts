import { createColumnHelper, type InitialTableState, type Row } from '@tanstack/vue-table'
import { actionsTable, expandButton, iconButton } from './generate-functions'

import {
  InitiativeTableRowName,
  InitiativeTableRowHealth,
  InitiativeTableRowActions,
  InitiativeTableRowConditions,
  InitiativeTableRowNotes,
  InitiativeTableRowInit,
  InitiativeTableRowAc,
  InitiativeTableRowDeathSaves,
  InitiativeTableRowConcentration,
  InitiativeTableRowModify,
} from '#components'

const columnHelper = createColumnHelper<InitiativeSheetRow>()

interface ColumnOptions {
  sheet: ComputedRef<InitiativeSheet | undefined>
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable>) => Promise<void>
  openQuickInitModal: () => void
}

export function generateColumns({ sheet, update, openQuickInitModal }: ColumnOptions) {
  const { t } = useI18n()

  return [
    columnHelper.display({
      enableGlobalFilter: false,
      id: 'index',
      cell: ({ row }) => h('div', { class: 'flex items-center gap-x-2 max-w-20' }, [
        h('span', { class: 'text-muted-foreground ml-1' }, row.index + 1),
        expandButton({
          content: t(`actions.${row.getIsExpanded() ? 'hide' : 'show'}`),
          expanded: row.getIsExpanded(),
          cb: () => row.toggleExpanded(),
        }),
      ]),
    }),
    columnHelper.accessor('name', {
      header: t('components.encounterTable.headers.name'),
      cell: ({ row }) => h(InitiativeTableRowName, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('initiative', {
      header: () => h('button', { class: 'flex items-center gap-x-2' }, [
        t('components.encounterTable.headers.init'),
        iconButton({
          icon: 'tabler:bolt-filled',
          content: t('components.encounterTable.quick'),
          color: 'warning',
          cb: () => openQuickInitModal(),
          disabled: !sheet.value?.rows.length,
          iconColor: 'text-warning',
        }),
      ]),
      cell: ({ row }) => h(InitiativeTableRowInit, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('health', {
      header: t('components.encounterTable.headers.health'),
      cell: ({ row }) => h(InitiativeTableRowHealth, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('ac', {
      header: t('components.encounterTable.headers.ac'),
      cell: ({ row }) => h(InitiativeTableRowAc, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      id: 'manage',
      header: t('components.encounterTable.headers.manage'),
      cell: ({ row }) => h(InitiativeTableRowActions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('conditions', {
      header: t('components.encounterTable.headers.conditions'),
      cell: ({ row }) => h(InitiativeTableRowConditions, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('note', {
      header: t('components.encounterTable.headers.note'),
      cell: ({ row }) => h(InitiativeTableRowNotes, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('deathSaves', {
      header: t('components.encounterTable.headers.deathSaves'),
      cell: ({ row }) => h(InitiativeTableRowDeathSaves, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.accessor('concentration', {
      header: t('components.encounterTable.headers.concentration'),
      cell: ({ row }) => h(InitiativeTableRowConcentration, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
    columnHelper.display({
      enableGlobalFilter: false,
      enableSorting: false,
      id: 'actions',
      cell: ({ row }) => h(InitiativeTableRowModify, {
        item: row.original,
        sheet: sheet.value,
        update,
      }),
    }),
  ]
}

export function expandedMarkup(row: Row<InitiativeSheetRow>) {
  return actionsTable(row.original)
}

export const initialState: InitialTableState = {}
