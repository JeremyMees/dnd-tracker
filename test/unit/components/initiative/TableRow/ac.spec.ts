import { mountSuspended } from '@nuxt/test-utils/runtime'
import { describe, expect, it, vi } from 'vitest'
import Ac from '~/components/initiative/TableRow/Ac.vue'
import { sheet } from '~~/test/unit/fixtures/initiative-sheet'

interface Props {
  item: InitiativeSheetRow
  sheet: InitiativeSheet | undefined
  update: (payload: Omit<Partial<InitiativeSheet>, NotUpdatable | 'campaign'>) => Promise<void>
}

const mockUpdate = vi.fn()

const props: Props = {
  item: sheet.rows[0]!,
  sheet,
  update: mockUpdate,
}

describe('Initiative table row ac', async () => {
  it('Should match snapshot', async () => {
    const component = await mountSuspended(Ac, { props })

    expect(component.html()).toMatchSnapshot()
  })

  it('Should display AC values correctly', async () => {
    const ac = 1
    const maxAc = 15
    const maxAcOld = 10
    const tempAc = 5

    const component = await mountSuspended(Ac, {
      props: {
        ...props,
        item: { ...props.item, ac, maxAc, maxAcOld, tempAc },
      },
    })

    expect(component.get('[data-test-ac]').text()).toBe(ac.toString())
    expect(component.get('[data-test-max]').text()).toContain(maxAc.toString())
    expect(component.get('[data-test-temp]').text()).toContain(tempAc.toString())
  })

  it('Should show destructive styling when AC is 0', async () => {
    const component = await mountSuspended(Ac, {
      props: {
        ...props,
        item: { ...props.item, ac: 0 },
      },
    })

    expect(component.get('[data-test-trigger]').classes()).toContain('bg-destructive/20')
    expect(component.get('[data-test-ac]').classes()).toContain('text-destructive')
  })

  it('Should show plus icon when AC is not defined', async () => {
    const component = await mountSuspended(Ac, {
      props: {
        ...props,
        item: { ...props.item, ac: undefined },
      },
    })

    expect(component.get('[data-test-empty]').isVisible()).toBeTruthy()
  })
})
