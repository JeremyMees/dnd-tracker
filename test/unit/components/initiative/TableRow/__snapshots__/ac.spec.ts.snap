// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Initiative table row ac > Should match snapshot 1`] = `
"<div><button data-test-trigger="" class="flex flex-col gap-y-1" id="reka-popover-trigger-v-0-0-0" type="button" aria-haspopup="dialog" aria-expanded="false" aria-controls="" data-state="closed">
    <div class="flex items-center gap-x-1"><span data-test-ac="" class="">1</span><span data-test-temp="" class="text-warning text-xs"> +5</span></div><span data-test-max="" class="text-2xs text-muted-foreground whitespace-nowrap">general.max: 19</span>
  </button>
  <!--teleport start-->
  <!--teleport end-->
</div>"
`;
